lib/AST/FineGrainedDependencyFormat.cpp

```diff
@@ -0,0 +1,482 @@
+    return NodeKind(nodeKind);
+  return None;
+}
+
+static llvm::Optional<DeclAspect> getDeclAspect(unsigned declAspect) {
+  if (declAspect < unsigned(DeclAspect::aspectCount))
+    return DeclAspect(declAspect);
+  return None;
+}
+
+bool Deserializer::readFineGrainedDependencyGraph(SourceFileDepGraph &g) {
+  using namespace record_block;
+
+  if (readSignature())
+    return true;
+
+  if (enterTopLevelBlock())
+    return true;
+
+  if (readMetadata())
+    return true;
+
+  SourceFileDepGraphNode *node = nullptr;
+  size_t sequenceNumber = 0;
+
+  while (!Cursor.AtEndOfStream()) {
+    auto entry = cantFail(Cursor.advance(), "Advance bitstream cursor");
+
+    if (entry.Kind == llvm::BitstreamEntry::EndBlock) {
+      Cursor.ReadBlockEnd();
+      assert(Cursor.GetCurrentBitNo() % CHAR_BIT == 0);
+      break;
+    }
+
+    if (entry.Kind != llvm::BitstreamEntry::Record)
+      llvm::report_fatal_error("Bad bitstream entry kind");
+
+    Scratch.clear();
+    unsigned recordID = cantFail(
+      Cursor.readRecord(entry.ID, Scratch, &BlobData),
+      "Read bitstream record");
+
+    switch (recordID) {
+    case METADATA: {
+      // METADATA must appear at the beginning and is handled by readMetadata().
+      llvm::report_fatal_error("Unexpected METADATA record");
+      break;
+    }
+
+    case SOURCE_FILE_DEP_GRAPH_NODE: {
+      unsigned nodeKindID, declAspectID, contextID, nameID, isProvides;
+      SourceFileDepGraphNodeLayout::readRecord(Scratch, nodeKindID, declAspectID,
+                                               contextID, nameID, isProvides);
+      node = new SourceFileDepGraphNode();
+      node->setSequenceNumber(sequenceNumber++);
+      g.addNode(node);
+
+      auto nodeKind = getNodeKind(nodeKindID);
+      if (!nodeKind)
+        llvm::report_fatal_error("Bad node kind");
+      auto declAspect = getDeclAspect(declAspectID);
+      if (!declAspect)
+        llvm::report_fatal_error("Bad decl aspect");
+      auto context = getIdentifier(contextID);
+      if (!context)
+        llvm::report_fatal_error("Bad context");
+      auto name = getIdentifier(nameID);
+      if (!name)
+        llvm::report_fatal_error("Bad identifier");
+
+      node->setKey(DependencyKey(*nodeKind, *declAspect, *context, *name));
+      if (isProvides)
+        node->setIsProvides();
+      break;
+    }
+
+    case FINGERPRINT_NODE: {
+      // FINGERPRINT_NODE must follow a SOURCE_FILE_DEP_GRAPH_NODE.
+      if (node == nullptr)
+        llvm::report_fatal_error("Unexpected FINGERPRINT_NODE record");
+
+      node->setFingerprint(BlobData);
+      break;
+    }
+
+    case DEPENDS_ON_DEFINITION_NODE: {
+      // DEPENDS_ON_DEFINITION_NODE must follow a SOURCE_FILE_DEP_GRAPH_NODE.
+      if (node == nullptr)
+        llvm::report_fatal_error("Unexpected DEPENDS_ON_DEFINITION_NODE record");
+
+      unsigned dependsOnDefID;
+      DependsOnDefNodeLayout::readRecord(Scratch, dependsOnDefID);
```

Comment:
(Meant to select this whole function, but too long)
*Not your fault at all,* but this bitstream stuff reads like COBOL--so verbose!
I'm skipping the rest of the file.

Reply:
Sure, it would be nice if, eg it could generate the reader and writer from a single declarative description. But that's not how it works today, I don't want to invent a way of doing that, and everyone else writes out a ton of boilerplate every time they use it too.

Reply:
Sigh... Someday, when it's all in Swift...